### PR TITLE
adds `scale` and `precision` methods to `Decimal` type

### PR DESCRIPTION
**Pull request by desaikd**
*Issue #363*

*Description of changes:*
This PR works on adding `scale` and `precsion` methods to `Decimal` type.

*List of changes:*
* adds `number_of_decimal_digits()` to `Magnitude` and `Coefficient`.
* adds `precision()` to `Decimal` types
* adds `scale()` to `Decimal` type

*Tests:*
* adds unit tests in `Decimal` type.

### CODE REVIEW

File: src/types/decimal.rs [hunk 1]
```diff
@@ -28,6 +28,23 @@ impl Decimal {
         }
     }
 
+    /// Returns scale of the Decimal value
+    /// A scale indicates the number of digits to the right of the decimal point.
```
[the remark this answers is not in the document]
Comment (desaikd) [reply src/types/decimal.rs:32]: I think you are right we should keep it similar to Java's `BigDecimal`.  I was taking a reference from [Ion schema specification](https://amzn.github.io/ion-schema/docs/spec.html#scale) for this implementation but that doesn't seem right here.

File: src/types/magnitude.rs [hunk 2]
```diff
@@ -67,6 +69,58 @@ impl Magnitude {
...
+    fn ten_to_the(pow: u64) -> BigUint {
+        if pow < 20 {
+            BigUint::from(10u64.pow(pow as u32))
+        } else {
+            let (half, rem) = pow.div_rem(&16);
+
+            let mut x = Magnitude::ten_to_the(half);
+
+            for _ in 0..4 {
+                x = &x * &x;
```
[the remark this answers is not in the document]
Comment (desaikd) [reply src/types/magnitude.rs:92]: Reference implementation for this: [BigDecimal's digits()](https://docs.rs/bigdecimal/latest/src/bigdecimal/lib.rs.html#97)

[the remark this answers is not in the document]
Comment (zslayton) [reply src/types/magnitude.rs:92]: I saw that, I just didn't understand what the reference implementation was doing. 😛

[the remark this answers is not in the document]
Comment (desaikd) [reply src/types/magnitude.rs]: Changed this method with https://github.com/amzn/ion-rust/pull/365/commits/17fce15ff316896e181353200997e7e30ecb9856

File: src/types/decimal.rs [hunk 3]
```diff
@@ -28,6 +28,23 @@ impl Decimal {
...
+    pub fn scale(&self) -> i64 {
+        if self.exponent > 0 {
+            return 0;
+        }
+        self.exponent.abs()
+    }
+
+    /// Returns the number of digits in the non-scaled integer representation of the decimal.
```
[the remark this answers is not in the document]
Comment (desaikd) [reply src/types/decimal.rs:39]: Changed the name to `precision`.

File: src/types/decimal.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (desaikd) [reply src/types/decimal.rs]: Changed this to return negated `exponent` as the `scale` value.

File: src/types/coefficient.rs [hunk 4]
```diff
@@ -39,6 +39,10 @@ impl Coefficient {
         &self.magnitude
     }
 
+    pub(crate) fn digits(&self) -> u64 {
```
[the remark this answers is not in the document]
Comment (desaikd) [reply src/types/coefficient.rs]: Added a comment for this with https://github.com/amzn/ion-rust/pull/365/commits/17fce15ff316896e181353200997e7e30ecb9856

File: src/types/coefficient.rs [hunk 5]
```diff
@@ -39,8 +39,9 @@ impl Coefficient {
         &self.magnitude
     }
 
-    pub(crate) fn digits(&self) -> u64 {
-        self.magnitude.digits()
+    /// Returns the number of digits in the non-scaled integer representation of the coefficient.
```
Comment (zslayton) [inline src/types/coefficient.rs:42]: Scaled vs non-scaled is a `Decimal` concept, so I think you should leave it out of this `Coefficient` doc comment.
```suggestion
    /// Returns the number of digits in the base-10 representation of the coefficient
```

File: src/types/decimal.rs [hunk 6]
```diff
@@ -29,20 +30,18 @@ impl Decimal {
     }
 
     /// Returns scale of the Decimal value
-    /// A scale indicates the number of digits to the right of the decimal point.
+    /// If zero or positive, a scale indicates the number of digits to the right of the decimal point.
+    /// If negative, the unscaled value of the number is multiplied by ten to the power of the negation of the scale. For example, a scale of -3 means the unscaled value is multiplied by 1000.
```
Comment (zslayton) [inline src/types/decimal.rs:34]: Nit: I suggest wrapping this line.
```suggestion
    /// If negative, the unscaled value of the number is multiplied by ten to the power of the negation of the scale.
    /// For example, a scale of -3 means the unscaled value is multiplied by 1000.
```

**Review by zslayton — APPROVED** [review]

**Pull request merged**